### PR TITLE
Ch3 issue70

### LINKED ISSUE #70: What are the platforms and their respective considerations required for collaborative ML development?
**Section Outline**
Tools for collaboration.
 - Communication tools.
 - Code versioning.
 - Model/Weight/Data versioning

 
**Definition of Done**
- [ ] Completed `3_Collaborative_Development_Platforms/collaborative_ml.md`
- [ ] Verified against [self-review checklist](https://github.com/aisingapore/ai-practitioner-handbook/blob/main/REVIEWING.md)
- [ ] Added section to [table of contents](https://github.com/aisingapore/ai-practitioner-handbook/blob/main/_toc.yml)
- [ ] Page builds and displays correctly on local machine

### PR DESCRIPTION
**Pull request by chuawjk**
Opening a new PR due to some workflow adjustments.

### CODE REVIEW

**Comment by chuawjk** [comment]
Hi all, please disregard any earlier PRs that you may have received for this section. I opened a new one due to some workflow adjustments.

I'd like to invite @kevinosl @najibninaba @ngkh2018 to review this contribution by @Unicorndy. Thanks again Dylan for helping out!

**Comment by najibninaba** [comment]
Would you want to consider mentioning the use of pull requests or merge requests and filing/resolving issues in GitHub or GitLab is also a form of collaboration in a development team?

File: book/3-collab-dev-platforms/collaborative_ml.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+Fundamentally, there are two ways that teams collaborate. <br>
+*Luke Marsden: When you boil it down to basics, there are really two fundamental modes of collaborating between different people doing work: synchronous and asynchronous.*
+
+*Synchronous*: Communications are scheduled, real-time interactions by phone, video, or in-person, sometimes interrupting one another for urgent tasks or simply asking a question. In terms of development, the upstream have to complete their tasks before the downstream team can continue with their work (Dependencies issues)
+
+*Asynchronous*: Communication happens on your own time, with a different team or people concurrently working on a different task. An efficient coping mechanism is required to resolve merged conflict. 
+
+With this in mind, we seek to introduce some tools and practices for a team to get started in the hope that stumbling blocks will be reduced to the minimum. The practices introduce here assume development using python as the primary language.
```
Comment (kevinosl) [inline book/3-collab-dev-platforms/collaborative_ml.md]: ```The practices introduced here... ```
Edited as above.

File: book/3-collab-dev-platforms/collaborative_ml.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+## Communication tools.
```
Comment (kevinosl) [inline book/3-collab-dev-platforms/collaborative_ml.md]: ```## Communication tools.```
Remove full stop in title to be consistent.

File: book/3-collab-dev-platforms/collaborative_ml.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+- [ Google Workspace](https://workspace.google.com/), a full suite of communication tools such as Google meet, shared Google calendar, and Gmail to see the complete list of participants and scheduled meetings. Other option includes the [Microsoft Team](https://www.microsoft.com/en-us/microsoft-teams/log-in).
+- [Miro](https://miro.com/) is an easy to use, intuitive, collobration tool that uses little effort to setup quickly. Other option includes [Coggle](https://coggle.it/).
+
+## Code versioning
+Jupyter notebook(.ipynb) although powerful has many limitations. Versioning is not one of them and is certainly not the best format to be deployed or go into production. Name and renaming files or folders with comments are also undesirable and difficult to keep track of. Instead, using [git](https://git-scm.com/) for distributed version control and tracking changes helps coordinate work among team members. Developing code in python script (.py) is highly preferred and using [Gitlab](https://about.gitlab.com/) or [Github](https://github.com/) to manage your code online with an advanced interface to resolve merged conflict that occurs frequently for asynchronous work.
```
Comment (kevinosl) [inline book/3-collab-dev-platforms/collaborative_ml.md]: ```Jupyter notebook(.ipynb) although powerful has many limitations. Versioning is not one of them...```
Do you mean to say versioning is one of them?
```, and it is certainly not the best format to be deployed or to go into production. Naming and renaming...```
Edited as above.

File: book/3-collab-dev-platforms/collaborative_ml.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+## Code versioning
+Jupyter notebook(.ipynb) although powerful has many limitations. Versioning is not one of them and is certainly not the best format to be deployed or go into production. Name and renaming files or folders with comments are also undesirable and difficult to keep track of. Instead, using [git](https://git-scm.com/) for distributed version control and tracking changes helps coordinate work among team members. Developing code in python script (.py) is highly preferred and using [Gitlab](https://about.gitlab.com/) or [Github](https://github.com/) to manage your code online with an advanced interface to resolve merged conflict that occurs frequently for asynchronous work. 
+
+Keeping a standardized environment for development keeps python package dependencies consistent, and together with Git, allows different members to easily switch between different machines such as local machines or machines in the cloud. One popular option that manages the environment is [Conda](https://docs.conda.io/en/latest/) (environment.yml) 
+
+More in-depth info can be found in 3-collab-dev-platforms/repo-structure-setup.md.
```
Comment (kevinosl) [inline book/3-collab-dev-platforms/collaborative_ml.md]: Should there be a link to the .md?

File: book/3-collab-dev-platforms/collaborative_ml.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+## Model/Weight/Data versioning
+Shared storage such as cloud storage ensures the team is assessing the correct dataset and having access to a shared transformed dataset cuts down data preprocessing/transformation time allowing the engineer to focus on modeling. Also, weights and datasets should not be push to the version control system, as the repository size might become too large for the local machine.
```
Comment (kevinosl) [inline book/3-collab-dev-platforms/collaborative_ml.md]: ```... should not be pushed...```

File: book/3-collab-dev-platforms/collaborative_ml.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+Having a system to keep track of the trained model and the corresponding dataset early on will prove to be fruitful, as more experiments will make it difficult to trace back. One good practice will be to save the checkpoints and the models together and keep track of the corresponding dataset used for the training. Recommended tools include [MLflow](https://mlflow.org/) and [Weights & Biases](https://wandb.ai/site) for tracking experiments and versioning models. [DVC](https://dvc.org/) and [neptune.ai](https://neptune.ai/) for Data Version Control. Take note even a shared spreadsheet keeping track of model/weights/dataset versioning will pay off eventually.
+
+It is also possible to set up cloud-based compute development environment where all collaborators can share the same environment and tools(VScode, Jupyter Lab) and shared storage. Some of the examples are Amazon EC2 Instance Types, Data Science Virtual Machine and Google Compute Engine
```
Comment (kevinosl) [inline book/3-collab-dev-platforms/collaborative_ml.md]: ```... to set up a cloud-based compute development environment where all collaborators can share the same environment and tools (VScode, Jupyter Lab) and shared storage. Some of...```
```... and Google Compute Engine.```
Edited as above.

**Comment by kevinosl** [comment]
@Unicorndy have reviewed. Thanks! Pardon the 'editorial' checks.
General comment: shall we standardise the use of Python (capitalized spelling) throughout the book?

File: book/3-collab-dev-platforms/collaborative_ml.md (hunk 4 above)
Comment (Unicorndy) [reply book/3-collab-dev-platforms/collaborative_ml.md]: Link included

File: book/3-collab-dev-platforms/collaborative_ml.md (hunk 1 above)
Comment (Unicorndy) [reply book/3-collab-dev-platforms/collaborative_ml.md]: Updated

File: book/3-collab-dev-platforms/collaborative_ml.md (hunk 2 above)
Comment (Unicorndy) [reply book/3-collab-dev-platforms/collaborative_ml.md]: Updated.

File: book/3-collab-dev-platforms/collaborative_ml.md (hunk 3 above)
Comment (Unicorndy) [reply book/3-collab-dev-platforms/collaborative_ml.md]: Updated

Jupyter notebook(.ipynb) although powerful has many limitations. `For example, it is impractical to use pure git for versioning Jupyter notebook,` and it is certainly not the best format to be deployed or go into production. `Naming and renaming`

File: book/3-collab-dev-platforms/collaborative_ml.md (hunk 5 above)
Comment (Unicorndy) [reply book/3-collab-dev-platforms/collaborative_ml.md]: Updated.

File: book/3-collab-dev-platforms/collaborative_ml.md (hunk 6 above)
Comment (Unicorndy) [reply book/3-collab-dev-platforms/collaborative_ml.md]: Updated

**Comment by Unicorndy** [comment]
> Would you want to consider mentioning the use of pull requests or merge requests and filing/resolving issues in GitHub or GitLab is also a form of collaboration in a development team?

Hi Najib, 
On line 25, I briefly talked about merge/pull requests. I've now expanded on this to better illustrate the point with examples.

**Pull request closed without merging**